tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+       For example, a DOI is registered but has no URL to resolve to.
+       Rare but theoretically possible.
+3. Something unspecified went wrong on the server end (HTTP status code 500).
+
+Omitting some bells and whistles from biblids, we can minimally wrap the API thus:
+
+```{r minimal-wrapper}
+# unexported helper
+# this is not tested, b/c the implementation could change
+verb_doi_handle <- function(doi, ...) {
+  httr::VERB(
+    url = "https://doi.org/",
+    # better to use curl::curl_escape() here
+    path = paste0("api/handles/", doi),
+    ...
+  )
+}
+
+# exported function
+# this is tested
+get_doi_handle <- function(doi, ...) {
+  resp <- verb_doi_handle(doi = doi, verb = "GET", ...)
+  httr::stop_for_status(resp) # covers outcomes 1 *and* 3 from above list
+  if (httr::content(resp)$responseCode == 200) {
+    # covers outcome 2b
+    stop("The handle exists, but the queried value(s) could not be found.")
+  }
+  # this must be outcome 2a
+  resp
+}
+
+# still missing: a way to disambiguate between outcomes 1 and 3
+# exported function, this is tested
+is_doi_found <- function(doi) {
+  # HEAD is more idiomatic, and marginally more efficient
+  resp <- verb_doi_handle(doi = doi, verb = "HEAD")
+  if (httr::status_code(resp) == 200) return(TRUE)
+  if (httr::status_code(resp) == 404) return(FALSE)
+  httr::stop_for_status(resp)
+}
+```
+
+A minimal interface between exported functions and the API is desirable, because it simplifies the below tests.
+`get_doi_handle()` would be the absolute minimum, but it alone cannot distinguish between API failure modes 1 from 3:
+`httr::stop_for_status()` could be thrown either because a DOI does not exist or because of some other HTTP error code.
+The differences between 1 and 3 are important for users and the downstream logic in the biblids package.
+We thus need an additional function at this interface, `is_doi_found()`, which checks explicity for failure mode 1.
+
+We can test most API failure modes of `get_doi_handle()` and `is_doi_found()` thus:
+
+```{r test-minimal-wrapper}
+library(testthat)
+skip_if_offline()
+# outcome 1
+test_that("Non-existent DOIs error out", {
+  expect_error(get_doi_handle("10.1000/I_DO_NOT_EXIST"))
+  expect_false(is_doi_found("10.1000/I_DO_NOT_EXIST"))
+})
+# outcome 2a
+test_that("Existing DOIs are found", {
+  testthat::local_edition(3)
+  # this snapshot will also capture some change in the API output
+  expect_snapshot_output(get_doi_handle("10.1038/nature14236"))
+  expect_true(is_doi_found("10.1038/nature14236"))
+})
+# outcome 2b
+test_that("Non-existing values error out", {
+  expect_error(
+    get_doi_handle(
+      doi = "10.1038/nature14236",
+      query = list(type = "I_DO_NOT_EXIST")
+    )
+  )
+})
+```
+
+One might think that these tests should be mocked, rather than running against the live API.
+However, because DOI.org is stable, performant and requires no authentication, the added complexity is not worth it in this case.
```

Comment:
What do the API maintainers say about this btw? 
```suggestion
However, because DOI.org is stable, performant and requires no authentication, the added complexity is not worth in my opinion it in this case.
```

Reply:
Their [documentation](https://www.doi.org/factsheets/DOIProxy.html#rest-api) is *very* sparse 🙂.
But I think it's backed by a proxy or CDN or some such thing, and it doesn't even *have* auth, so I'd think these relatively few requests (dozends) per test run are ok.

Reply:
the "in my opinion" seems a bit superfluous, no?